# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [26]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   среднее             1   
8         2    6929.865299         35    высшее             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы  
5               покупка жилья                     B  операции с недвижимостью  
6           операции с жильем                     B  операции с недвижимостью  
7                 образование                     C     получение образования  
8       на проведение свадьбы                     C        проведение свадьбы  
9     покупка жилья для семьи                     C  операции с недвижимостью

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
def data_pivot(category):
    return data.pivot_table(['debt'], index=[category], aggfunc = ['count', 'sum', 'mean'])\
.rename(columns={'count':'Всего кредитополучателей', 'sum':'Всего должников', 'mean':'Доля должников'}).droplevel(1,axis=1)
data_pivot(data['children'])

Всего кредитополучателей  Всего должников  Доля должников
children                                                           
0                            14091             1063        0.075438
1                             4808              444        0.092346
2                             2052              194        0.094542
3                              330               27        0.081818
4                               41                4        0.097561
5                                9                0        0.000000

**Вывод:** Люди без детей реже имеют дебиторскую задолженность, лишь 7.54%. Люди с тремя детьми немного чаще, 8.18%. Про людей с четырьмя и с пятью детьми сложно сказать, так как из всей выборки только у 9 людей 5 детей, у 41 человека 4 детей, а в нашей выборке это очень маленькие значения для полноценной оценки.
Следовательно, самые надежные плательщики - люди без детей.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_pivot(data['family_status'])

Всего кредитополучателей  Всего должников  \
family_status                                                      
Не женат / не замужем                      2796              273   
в разводе                                  1189               84   
вдовец / вдова                              951               63   
гражданский брак                           4134              385   
женат / замужем                           12261              927   

                       Доля должников  
family_status                          
Не женат / не замужем        0.097639  
в разводе                    0.070648  
вдовец / вдова               0.066246  
гражданский брак             0.093130  
женат / замужем              0.075606

**Вывод:** Самые надёжные заёмщики - вдовы/вдовцы и люди в разводе, самые ненадёжные - не женатые/ не замужем

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data_pivot(data['total_income_category'])

Всего кредитополучателей  Всего должников  \
total_income_category                                              
A                                            25                2   
B                                          5014              354   
C                                         15921             1353   
D                                           349               21   
E                                            22                2   

                       Доля должников  
total_income_category                  
A                            0.080000  
B                            0.070602  
C                            0.084982  
D                            0.060172  
E                            0.090909

**Вывод:** Самые надёжные заёмщики - люди из B категории, т.е. с доходом от 200001 до 1000000, чуть менее надёжные - из C категории, т.е. с доходом от 50001 до 200000, по остальным категориям судить сложно из-за недостатка данных

#### Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_pivot(data['purpose_category'])

Всего кредитополучателей  Всего должников  \
purpose_category                                                      
операции с автомобилем                        4279              400   
операции с недвижимостью                     10751              780   
получение образования                         3988              369   
проведение свадьбы                            2313              183   

                          Доля должников  
purpose_category                          
операции с автомобилем          0.093480  
операции с недвижимостью        0.072551  
получение образования           0.092528  
проведение свадьбы              0.079118

**Вывод:** Меньше всего задолженностей по кредитам с целью операции с недвижимостью, больше всего - по кредитам с целью операции с автомобилем

#### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Возможен человеческий фактор, например, клиент не предоставил подтверждающих документов, возможно их и не требовали (например, может быть не требуются какие-то данные если клиент имеет хорошую кредитную историю в этом банке или сумма кредита небольшая), или могла быть ошибка сотрудника, заполнявшего данные. Также возможен технический фактор - произошел какой-то сбой, возникла ошибка во время переноса, объединения, выгрузки или загрузки данных.

#### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что медианное значение позволяет сгладить влияние значений, выделяющихся из общей выборки.

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

В зависимости от вероятности возврата клиентом кредита в срок можно регулировать процентную ставку кредитного продукта, а также выдавать отказ потенциальному заёмщику в случае его неблагонадёжности.
Проведён анализ данных со статистикой платёжеспособности клиентов, удалены пропуски в данных, обработаны аномальные значения,  обработаны дубликаты. Также быда сделана категоризация данных и исследована зависимость между количеством детей, семейным положением, целью получения кредита, уровнем дохода и возвратом кредита в срок.
Выяснили, что вероятность возврата кредита в срок меняется следующим образом:
- вероятность задолженности низкая (7.54%), если у заёмщика нет детей, и выше, если у заёмщика 1 или 2 детей (9.23% и 9.45% соответственно)
- самые надёжные заёмщики по семейному положению - вдовы/вдовцы (вероятность задолженности 6.62%) и люди в разводе (вероятность задолженности 7.06%), самые ненадёжные - не женатые/ не замужем (вероятность задолженности 9.76%)
- самые надёжные заёмщики по уровню дохода - люди с доходом от 200001 до 1000000 (вероятность задолженности 7.06%), чуть менее надёжные - с доходом от 50001 до 200000 (вероятность задолженности 8.49%), по остальным категориям судить сложно из-за недостатка данных
- меньше всего задолженностей по кредитам с целью операции с недвижимостью (7.25%), больше всего - по кредитам с целью операции с автомобилем (9.34%)

Таким образом, самый надёжный заёмщик - человек без детей, вдовец/вдова или в разводе, с доходом от 200001 до 1000000, который берёт кредит для операции с недвижимостью.